### ISSUE TITLE
Socket support for TCP_QUICKACK

### ISSUE THREAD

**Issue by fujiy** [dotnet/corefx#29917]
I'm not sure if Windows support this at socket level as Linux does. But delayed ack does not works well when the remote endpoint is using Nagle Algorithm (TCP_NODELAY = false).

Reference: https://blogs.technet.microsoft.com/nettracer/2013/01/05/tcp-delayed-ack-combined-with-nagle-algorithm-can-badly-impact-communication-performance/

**Comment by karelz** [comment]
I couldn't find existing option on Windows sockets APIs either.
@davidsh should we ping Windows folks to confirm?

cc @geoffkizer

**Comment by davidsh** [comment]
>@davidsh should we ping Windows folks to confirm?

Yes @rmkerr could follow up with the usual TCP team that he already has worked with.

**Comment by rmkerr** [comment]
I'll reach out to the TCP team to confirm. In the meantime, I recommend checking out the documentation [here](https://support.microsoft.com/en-us/help/214397/design-issues-sending-small-data-segments-over-tcp-with-winsock) on managing some of the issues with the Nagle algorithm and delayed ack.

**Comment by fujiy** [comment]
@rmkerr, I read that article, one of the issues I intend to prevent is the problem with Nagle algorithm

**Comment by fujiy** [comment]
@rmkerr, how did you find the constant 12? Would it work on Linux or we need to map to another value in https://github.com/dotnet/corefx/blob/18cac5a386bdf7dba8585a78c047d77345379a71/src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs#L1092 or https://github.com/dotnet/corefx/blob/21f15c1422643e3f759f13d72867759fd797ca60/src/Native/Unix/System.Native/pal_networking.c#L1760 ?

**Comment by rmkerr** [comment]
This issue actually needs to be updated -- the API is slightly different than expected on Windows. Rather than using the standard `setsocketopt` interface, Windows allows for more granular control over ack frequency via the `WSAIoctl` control code `SIO_TCP_SET_ACK_FREQUENCY`. That means that we can add the ability to disable delayed ack on all platforms, but the implementation may need to be platform specific. I think the implementation details need more discussion.

**Comment by alfredmyers** [comment]
While researching on the topic, I found an issue over on the Kestrel repo.

The folks over there had an issue about it but it was closed  https://github.com/aspnet/KestrelHttpServer/issues/480

**Comment by juliusfriedman** [comment]
I have some methods related to this however by other names, e.g

I believe the name "CongestionAlgorithm" is appropriate and the api allows a broader range of options:

      #region CongestionAlgorithm

        const int TcpCongestionAlgorithm = 12;

        static readonly System.Net.Sockets.SocketOptionName TcpCongestionAlgorithmOption = (System.Net.Sockets.SocketOptionName)TcpCongestionAlgorithm;

        public static void SetTcpCongestionAlgorithm(System.Net.Sockets.Socket socket, int value = 1)
        {
            SetTcpOption(socket, TcpCongestionAlgorithmOption, value);
        }

        public static void DisableTcpCongestionAlgorithm(System.Net.Sockets.Socket socket)
        {
            SetTcpCongestionAlgorithm(socket, 0);
        }

        public static void EnableTcpCongestionAlgorithm(System.Net.Sockets.Socket socket, int algorithmType = 1)
        {
            SetTcpCongestionAlgorithm(socket, algorithmType);
        }

        #endregion

Also for Retransmission:

    #region Retransmission

        static System.Net.Sockets.SocketOptionName TcpMaximumRetransmissionOptionName = (System.Net.Sockets.SocketOptionName)(Common.Extensions.OperatingSystemExtensions.IsWindows ? 5 : 18);

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static void SetMaximumTcpRetransmissionTime(System.Net.Sockets.Socket socket, int amountInSeconds = 3)
        {
            //On windows this is TCP_MAXRT elsewhere USER_TIMEOUT

            //Mono checks the options and will not call setsocketopt if the name is not known to the Mono Runtime.
            //A work around would be to either define the call for get and set socketopt in this library or call the SetSocketOption_internal method for mono.
            SetTcpOption(socket, TcpMaximumRetransmissionOptionName, amountInSeconds);
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static void DisableTcpRetransmissions(System.Net.Sockets.Socket socket)
        {
            SetMaximumTcpRetransmissionTime(socket, 0);
        }

        [System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.AggressiveInlining)]
        public static void EnableTcpRetransmissions(System.Net.Sockets.Socket socket)
        {
            SetMaximumTcpRetransmissionTime(socket);
        }

        public static int GetMaximumTcpRetransmissionTime(System.Net.Sockets.Socket socket)
        {
            int len = Common.Binary.BytesPerInteger;

            //Todo, static local allocation
            byte[] value = socket.GetSocketOption(System.Net.Sockets.SocketOptionLevel.Tcp, TcpMaximumRetransmissionOptionName, len);

            if (Common.Extensions.Array.ArrayExtensions.IsNullOrEmpty(value, out len) || len < Common.Binary.BytesPerInteger) return -1;

            return Common.Binary.Read32(value, Common.Binary.BytesPerInteger, BitConverter.IsLittleEndian);

        }

        #endregion

And for others (most already with baked in checked for OS) see(https://github.com/juliusfriedman/net7mma/blob/master/Common/Extensions/SocketExtensions.cs):

NoSynRetries, Timestamp, OffloadPreference, DelayFinAck, GetMaximumTransmittableUnit, Urgency and last but not least MaximumSegmentSize.

**Comment by stephentoub** [comment]
> Depending on the results of that discussion we may need to go through API review again.

Removing api-approved until this is resolved.
